p4-16/psa/PSA.mdk

```diff
@@ -950,41 +940,33 @@ used in the second pass. See Section [#sec-open-metadata] for the
 discussion.
 
 A PSA implementation provides a configuration bit `resubmit` to the
-PRE to enable the resubmission mechanism. If `resubmit` is set to
-true, the resubmission mechanism resends the original packet with the
-optional attached metadata. If `resubmit` is set to false, the
-resubmission mechanism is disabled and the original packet is not
-resubmitted even if the `add_metadata` method is invoked in the
-deparser.
-
-```
-[INCLUDE=psa.p4:Resubmit_extern]
-
-```
+PRE to enable the resubmission mechanism. If true, the original packet
+is resubmitted with the optional resubmit metadata. If false, the
+resubmission mechanism is disabled and the value assigned to the
+`resubmit_meta` is ignored.
 
 ## Packet Recirculation {#sec-recirculate}
 
-Packet recirculation sends the deparsed packet at the end of egress
-pipeline to the ingress parser. Unlike the previous two mechanisms,
-recirculation does not replicate packets. The modified packet itself
-is sent back to the ingress pipeline for further processing.
-
-In PSA, recirculation is controlled by the `recirculate` bit in the
-egress pipeline. If the `recirculate` bit is set to true, the modified
-packet at the end of the pipeline is sent to a special recirculation
-port. The recirculation port overrides the value in `egress_port`.
-
-```
-[INCLUDE=psa.p4:Recirculate_extern]
+Packet recirculation is a mechanism to repeat both ingress and egress
+processing on a packet.
 
-```
+Packet recirculation happens at the end of the egress pipeline. When a
+packet is recirculated, the packet finishes both ingress and egress
+processing and re-enters the ingress parser after being deparsed by
+the egress deparser. The `ingress_port` of the recirculated packet is
+set to `RECIRCULATE`. The `packet_path` of the recirculated packet is
+changed to `RECIRCULATE`.
 
-Users can leverage the existing `emit` methods in `packet_out` to
-attach additional metadata to the recirculated packet. Upon receiving
-the recirculated packet in ingress parser, the `packet_path` of the
-recirculated packet is set to `RECIRCULATE`. It is the user's
-responsibility to parse the recirculated packet correctly.
+~ TBD
+Whether use two different enums for recirc port and recirc path
+~
 
+The `recirculate` flag specifies whether a packet should be
+recirculated. If true, the processed packet should be recirculated at
+the end of the egress pipeline. The `egress_port` must be set to the
+dedicated port for recirculation in the pipeline. Some target may
+restrict the `recirculate` flag and `egress_port` to be set in the
```

Comment:
Nothing else in the current draft document says anything about having to assign a value to `egress_port` in order to recirculate a packet.  It seems easier to only require the P4 program to assign true to the `recirculate` flag, and then they don't have to do anything else.

Also, the field `egress_port` is an in parameter to the egress control block, so it is illegal in P4_16 to attempt to assign a value to it.

Reply:
You mention extra metadata that can be associated with a resubmitted packet, but not for a recirculated packet.  It would be good to mention for both or neither -- for both seems a good choice.